Multiprocess build support

ryanofsky opened the pull request:
This PR is part of the [process separation project](https://github.com/bitcoin/bitcoin/projects/10).

---

This PR consists of build changes only. It adds an `--enable-multiprocess` autoconf option (off by default and marked experimental), that builds new `bitcoin-node` and `bitcoin-gui` binaries. These currently function the same as existing `bitcoind` and `bitcoin-qt` binaries, but are extended in #10102 with IPC features to execute node, wallet, and gui functions in separate processes.

In addition to adding the `--enable-multiprocess` config flag, it also adds a depends package and autoconf rules to build with the [libmultiprocess](https://github.com/chaincodelabs/libmultiprocess) library, and it adds new travis configuration to exercise the build code and run functional tests with the new binaries.

The changes in this PR were originally part of #10102 but were moved into #16367 to be able to develop and review the multiprocess build changes independently of the code changes. #16367 was briefly merged and then reverted in #18588. Only change since #16367 has been dropping the `native_boost.mk` depends package which was pointed out to be no longer necessary in https://github.com/bitcoin/bitcoin/pull/16367#issuecomment-596484337 and https://github.com/bitcoin/bitcoin/pull/18588#pullrequestreview-391765649

ryanofsky committed:
multiprocess: add multiprocess travis configuration

ryanofsky committed:
depends: add MULTIPROCESS depends option

Builds required packages and passes --enable-multiprocess option to bitcoin build

ryanofsky committed:
build: multiprocess autotools changes

autoconf and automake changes to support multiprocess gui/node/wallet execution.

This adds a new --enable-multiprocess flag, and build configuration code to
detect libraries needed for multiprocess support. The --enable-multiprocess
flag builds new bitcoin-node and bitcoin-gui executables, which are updated in
https://github.com/bitcoin/bitcoin/pull/10102 to communicate across processes.
But for now they are functionally equivalent to existing bitcoind and
bitcoin-qt executables.

MarcoFalke commented:
re-ACK c8672dfeb7443d83fbae02c1b1cd867f83b972b1, only change since my last review is dropping boost and small doc fixup. This review does not cover libmultiprocess 🚕

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK c8672dfeb7443d83fbae02c1b1cd867f83b972b1, only change since my last review is dropping boost and small doc fixup. This review does not cover libmultiprocess 🚕
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgnPQwAntHVvhb+Ic5V+GQI8wMKrY52J9CGmWu9dLo/0dH9zC3oO+dyFk4BRcNT
HZ5Hai5yy7+OxfWvqf88isAqlUfNa7z6jmT/YZfa3OviLufzS00kazMn1CFhAOvJ
O2as6oAZUifG6YrQmtSCTw1fN79oCq/NP5VkufNauSvHDtRhYIiK7p3kZqo4Jm8w
hs5WzlEqcRpxpkbbLYXRh4Be6Q7UL5cRGhYz9U8V79uo0Cp3klo+onXwofrYc/jq
tBMVAS5Qu7tcLEAIsr/NLlp+OSjy1/gkeM6vlR3gmvzj86b7gKnHpBQeQZI1PcNo
eZ/cN2kDgB5/fLIT+KX9e25qiSJm0XCAmN7fI64oTI85PRpQcVALEkf+jJmgGh8r
urLfBT6WUesJMHLcQLrOrdvX2aFMiLE/vSjuxpRqKJhSGvFLCg86R6t0SflXB6WF
FTp9VW76Ofi8ZWcA0+cU1bHjF2GUP33TddGE+O9posORfibm2FQ4n43gfp90izIE
/YYDPbeG
=IWHw
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `5ffc2ebe9acc3666c3cfa34fc843d6447defa592787071c5cc86bf728ace8075  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401085ffc2ebe9acc3666c3cfa34fc843d6447defa592787071c5cc86bf728ace8075f010a0950a15693905af24a6f5bea0bab69f08fff01088d726e9abcbf27eabccc3ec95b9d0e108f1045e98c68ef008aa9925ad2f01b5d20083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010911661598ded770dd1f56597adddc69208f1045e98c68ff0081f84f853f04eae680083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010f87517bd823731173f066bb97e39eaff08f1045e98c68ff008663f79fc67ae06230083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010fec898ea3503f1a10345aec93d3fa89508f1045e98c68ff008b35584d1f649b6750083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

fanquake commented on depends/builders/darwin.mk:
```diff
@@ -1,5 +1,5 @@
-build_darwin_CC:=$(shell xcrun -f clang)
-build_darwin_CXX:=$(shell xcrun -f clang++)
+build_darwin_CC:=clang
+build_darwin_CXX:=clang++
```
Can you clarify why this is needed? I did look at the commit comments about passing `--sysroot` and missing includes, however from what I can tell, there should be no change in behaviour with this diff. I did run a few compilation tests and didn't see any obvious difference in compile flags/includes etc. 

My assumption is that the problems seen (https://github.com/chaincodelabs/libmultiprocess/issues/5, https://github.com/bitcoin/bitcoin/pull/16367#issuecomment-592093081) might have been due to an issue with the development environments i.e missing SDK headers, not installing the CommandLineTools etc.

Mostly curious because if this was an issue it seems like it shouldn't be multiprocess specific? It also seems strange to drop the `xcrun` call only for clang/clang++, and not for the rest of the tools.

fanquake reviewed (commented):
Thanks for reopening and the responses [here](https://github.com/bitcoin/bitcoin/pull/18588#pullrequestreview-391765649). I also read through the IRC logs  from last night.

Few comments other than on the code. Nothing blocking while this is all still experimental.

Cross-compiling for macOS (on Debian 10.3) with `MULTIPROCESS=1` currently fails while building `capnp`:
```bash
make -C depends/ HOST=x86_64-apple-darwin16 MULTIPROCESS=1
...
libtool: link: clang++ -target x86_64-apple-darwin16 -mmacosx-version-min=10.12 --sysroot /bitcoin/depends/SDKs/MacOSX10.14.sdk -stdlib=libc++ -I./src -I./src -DKJ_HEADER_WARNINGS -DCAPNP_HEADER_WARNINGS -DCAPNP_INCLUDE_DIR=\"/bitcoin/depends/x86_64-apple-darwin16/include\" -D_THREAD_SAFE -pipe -O2 -D_THREAD_SAFE -D_THREAD_SAFE -o .libs/capnp src/capnp/compiler/module-loader.o src/capnp/compiler/capnp.o -Wl,-bind_at_load  -L/bitcoin/depends/x86_64-apple-darwin16/lib ./.libs/libcapnpc.dylib ./.libs/libcapnp-json.dylib /bitcoin/depends/work/build/x86_64-apple-darwin16/capnp/0.7.0-87ee694a224/.libs/libcapnp.dylib ./.libs/libcapnp.dylib /bitcoin/depends/work/build/x86_64-apple-darwin16/capnp/0.7.0-87ee694a224/.libs/libkj.dylib ./.libs/libkj.dylib
echo capnp capnpc-c++ src/capnp/test.capnp src/capnp/test-import.capnp src/capnp/test-import2.capnp src/capnp/compat/json-test.capnp | (read CAPNP CAPNPC_CXX SOURCES && ./$CAPNP compile --src-prefix=./src -o./$CAPNPC_CXX:src -I./src $SOURCES)
./capnp: line 117: /bitcoin/depends/work/build/x86_64-apple-darwin16/capnp/0.7.0-87ee694a224/.libs/capnp: cannot execute binary file: Exec format error
./capnp: line 117: /bitcoin/depends/work/build/x86_64-apple-darwin16/capnp/0.7.0-87ee694a224/.libs/capnp: Success
make[1]: *** [Makefile:4326: test_capnpc_middleman] Error 126
make[1]: Leaving directory '/bitcoin/depends/work/build/x86_64-apple-darwin16/capnp/0.7.0-87ee694a224'
make: *** [funcs.mk:254: /bitcoin/depends/work/build/x86_64-apple-darwin16/capnp/0.7.0-87ee694a224/./.stamp_built] Error 2
make: Leaving directory '/bitcoin/depends'
```

Would you prefer these issues opened here or the libmultiprocess repo?

Can we pass `--with-openssl=no` (+ any other relevant flags) to the capnp configure? I assume we aren't using any parts of it that require OpenSSL, and we might as well opt out of any [autodetection](https://github.com/capnproto/capnproto/blob/master/c%2B%2B/configure.ac#L148) and/or additional compilation.

Also, I didn't notice before, but just wanted to note that building capnp currently [requires gnu extensions](https://github.com/capnproto/capnproto/blob/master/c%2B%2B/CMakeLists.txt#L69). This is at odds with our current [compiler requirements](https://github.com/bitcoin/bitcoin/pull/18088).

MarcoFalke commented on src/Makefile.am:
```diff
@@ -546,22 +550,22 @@ libbitcoin_cli_a_SOURCES = \
-bitcoind_CXXFLAGS = $(AM_CXXFLAGS) $(PIE_FLAGS)
-bitcoind_LDFLAGS = $(RELDFLAGS) $(AM_LDFLAGS) $(LIBTOOL_APP_LDFLAGS)
+# bitcoind & bitcoin-node binaries #
+bitcoin_common_sources = bitcoind.cpp
```
Sorry to be bikeshedding this again, but what about naming this `node_sources` or `daemon_sources`? As of now it sounds confusingly similar to libbitcoin_common, which is an utility library for all Bitcoin Core util binaries

ryanofsky reviewed (commented):
Updated c8672dfeb7443d83fbae02c1b1cd867f83b972b1 -> f80a689013710a695370e207db0048373d252043 ([`pr/ipc-build2.1`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build2.1) -> [`pr/ipc-build2.2`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build2.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-build2.1..pr/ipc-build2.2)) with x86_64-apple-darwin16 fixes and suggested automake renames

re: https://github.com/bitcoin/bitcoin/pull/18677#pullrequestreview-395075571

> Cross-compiling for macOS (on Debian 10.3) with `MULTIPROCESS=1` currently fails while building `capnp`:
> 
> ```shell
> make -C depends/ HOST=x86_64-apple-darwin16 MULTIPROCESS=1
> ...
> ```

Good catch, I hadn't tested this configuration. I fixed this and more problems with it in the latest push.

> Would you prefer these issues opened here or the libmultiprocess repo?

This was a depends system issue so it made sense here, but in general it can be hard to pin down where issues arise so either place should be fine

> Can we pass `--with-openssl=no` (+ any other relevant flags) to the capnp configure? I assume we aren't using any parts of it that require OpenSSL, and we might as well opt out of any [autodetection](https://github.com/capnproto/capnproto/blob/master/c%2B%2B/configure.ac#L148) and/or additional compilation.

I think you'd have to write an openssl depends package and add it to `$(package)_dependencies` in capnp.mk for there to be anything to detect. But maybe requesting to be built without openssl when openssl isn't available would be a good way to demonstrate not liking openssl

> Also, I didn't notice before, but just wanted to note that building capnp currently [requires gnu extensions](https://github.com/capnproto/capnproto/blob/master/c%2B%2B/CMakeLists.txt#L69). This is at odds with our current [compiler requirements](https://github.com/bitcoin/bitcoin/pull/18088).

I don't think gnu extensions are required, and I don't see any problem building without them (following regular build steps but passing CXXFLAGS=-std=c++14 to configure). Digging into the history of this check it looks like it was added to work around a cygwin bug where snprintf wasn't declared with -std=c++0x, but was with -std=gnu++0x

https://groups.google.com/d/msg/capnproto/l46nHozrhCY/vJTHMkJC-ZEJ
https://groups.google.com/d/msg/capnproto/l46nHozrhCY/UG__SqiVbywJ
https://stackoverflow.com/questions/20149633/how-to-use-snprintf-in-g-std-c11-version-4-8-2

luke-jr commented:
I thought we were going to pivot this to just add a `-process=node/wallet/gui` option to the same `bitcoin-qt` binary?

ryanofsky commented:
Updated f80a689013710a695370e207db0048373d252043 -> fba09de775f1eff68d2f72650bccc418cbe87cf8 ([`pr/ipc-build2.2`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build2.2) -> [`pr/ipc-build2.3`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build2.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-build2.2..pr/ipc-build2.3)) to try to fix travis error https://travis-ci.org/github/bitcoin/bitcoin/jobs/677378707 missing pthread.h
Updated fba09de775f1eff68d2f72650bccc418cbe87cf8 -> 474f4f1d02267c158001c5af0a558d24ef5f971d ([`pr/ipc-build2.3`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build2.3) -> [`pr/ipc-build2.4`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build2.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-build2.3..pr/ipc-build2.4)) fixing travis error and moving mac os builder workaround to #18743

---

re: https://github.com/bitcoin/bitcoin/pull/18677#issuecomment-619182425

> I thought we were going to pivot this 

I think the current naming is better, but I pointed out in the IRC meeting it is easy to change. If it's important, would suggest opening a followup PR and considering this as usability change to discuss ideally after having some experience using the feature

hebasto commented on configure.ac:
```diff
@@ -1369,6 +1387,50 @@ AM_CONDITIONAL([EMBEDDED_UNIVALUE],[test x$need_bundled_univalue = xyes])
+  build_multiprocess=no
+fi
+
+AM_CONDITIONAL([BUILD_MULTIPROCESS],[test "x$build_multiprocess" = xyes])
```
The `BUILD_MULTIPROCESS` conditional seems not used in `Makefiles`.

hebasto commented on src/Makefile.am:
```diff
@@ -614,29 +630,14 @@ bitcoin_tx_LDADD += $(BOOST_LIBS)
+bitcoin_wallet_CPPFLAGS = $(bitcoin_bin_cppflags)
+bitcoin_wallet_CXXFLAGS = $(bitcoin_bin_cxxflags)
+bitcoin_wallet_LDFLAGS = $(bitcoin_bin_ldflags)
+bitcoin_wallet_LDADD = $(LIBBITCOIN_WALLET_TOOL) $(bitcoin_bin_ldadd)
```
With this PR `bitcoin_wallet_LDADD` contains `$(LIBBITCOIN_SERVER)` comparing to the master.
Is this needed?

hebasto commented:
7c99f2c45e23ae1ce8c5b539d14109497ac4322c:
Maybe add fallback to default value for the `TEST_RUNNER_ENV` variable in the `00_setup_env.sh` script for consistency?

ryanofsky reviewed (commented):
Updated 474f4f1d02267c158001c5af0a558d24ef5f971d -> abf920e460b72f43daef1b006cb13c1f1ee2e256 ([`pr/ipc-build2.4`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build2.4) -> [`pr/ipc-build2.5`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build2.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-build2.4..pr/ipc-build2.5)) with suggestions

Thanks for review!

re: https://github.com/bitcoin/bitcoin/pull/18677#issuecomment-622745712

> [7c99f2c](https://github.com/bitcoin/bitcoin/commit/7c99f2c45e23ae1ce8c5b539d14109497ac4322c):
> Maybe add fallback to default value for the `TEST_RUNNER_ENV` variable in the `00_setup_env.sh` script for consistency?

Added

ryanofsky commented:
Rebased abf920e460b72f43daef1b006cb13c1f1ee2e256 -> e1a8c361c26ae3251b21f505f9538336fdb14155 ([`pr/ipc-build2.5`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build2.5) -> [`pr/ipc-build2.6`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build2.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-build2.5-rebase..pr/ipc-build2.6)) due to conflict with #18899

hebasto commented on configure.ac:
```diff
@@ -1427,6 +1445,50 @@ AM_CONDITIONAL([EMBEDDED_UNIVALUE],[test x$need_bundled_univalue = xyes])
+elif test "x$with_libmultiprocess" != xno; then
+  AC_MSG_ERROR([--with-libmultiprocess=$with_libmultiprocess value is not yes, auto, or no])
+fi
+AC_SUBST(LIBMULTIPROCESS_CFLAGS)
+AC_SUBST(LIBMULTIPROCESS_LIBS)
```
In this PR these variables are not used in `Makefile`s.

But in #10102 one can see:
https://github.com/bitcoin/bitcoin/blob/aa38570fa5f1ceee5883a8170da60206051e05fd/src/Makefile.am#L597
https://github.com/bitcoin/bitcoin/blob/aa38570fa5f1ceee5883a8170da60206051e05fd/src/Makefile.am#L649

Is it intended?

hebasto replied:
> If it's a concern about clarity, is there something beyond these two lines of code that is confusing?

Yes, it is about clarity. Thank you for your patient clarification regarding future changes.

ryanofsky replied:
> Yes, it is about clarity. Thank you for your patient clarification regarding future changes.

Thank you for reviewing and testing and finding things to fix! I'd very much like to add comments to the build files to document things which are surprising. But I can't do that without knowing which specific things are surprising, so the more context provided in review comments, the better!

hebasto commented:
Testing e1a8c361c26ae3251b21f505f9538336fdb14155 on Linux Mint 19.3 (x86_84):
```
$ make -C depends NO_QT=1 MULTIPROCESS=1
...
Extracting libmultiprocess...
/home/hebasto/GitHub/bitcoin/depends/sources/5741d750a04e644a03336090d8979c6d033e32c0.tar.gz: OK
Preprocessing libmultiprocess...
Configuring libmultiprocess...
-- The CXX compiler identification is GNU 7.5.0
-- Check for working CXX compiler: /usr/bin/g++
-- Check for working CXX compiler: /usr/bin/g++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") 
-- Checking for module 'capnp'
--   Found capnp, version 0.7.0
-- Checking for module 'capnp-rpc'
--   Found capnp-rpc, version 0.7.0
-- Checking for module 'capnp-json'
--   Found capnp-json, version 0.7.0
-- Checking for module 'kj'
--   Found kj, version 0.7.0
-- Checking for module 'kj-async'
--   Found kj-async, version 0.7.0
-- Checking for module 'kj-http'
--   Found kj-http, version 0.7.0
-- Checking for module 'kj-test'
--   Found kj-test, version 0.7.0
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - not found
CMake Error at /usr/share/cmake-3.10/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find Threads (missing: Threads_FOUND)
Call Stack (most recent call first):
  /usr/share/cmake-3.10/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.10/Modules/FindThreads.cmake:205 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  CMakeLists.txt:11 (find_package)


-- Configuring incomplete, errors occurred!
See also "/home/hebasto/GitHub/bitcoin/depends/work/build/x86_64-pc-linux-gnu/libmultiprocess/5741d750a04e644a03336090d8979c6d033e32c0-037d09711be/CMakeFiles/CMakeOutput.log".
See also "/home/hebasto/GitHub/bitcoin/depends/work/build/x86_64-pc-linux-gnu/libmultiprocess/5741d750a04e644a03336090d8979c6d033e32c0-037d09711be/CMakeFiles/CMakeError.log".
funcs.mk:258: recipe for target '/home/hebasto/GitHub/bitcoin/depends/work/build/x86_64-pc-linux-gnu/libmultiprocess/5741d750a04e644a03336090d8979c6d033e32c0-037d09711be/./.stamp_configured' failed
make: *** [/home/hebasto/GitHub/bitcoin/depends/work/build/x86_64-pc-linux-gnu/libmultiprocess/5741d750a04e644a03336090d8979c6d033e32c0-037d09711be/./.stamp_configured] Error 1
make: Leaving directory '/home/hebasto/GitHub/bitcoin/depends'
```

ryanofsky reviewed (commented):
re: https://github.com/bitcoin/bitcoin/pull/18677#issuecomment-625674673

> Testing [e1a8c36](https://github.com/bitcoin/bitcoin/commit/e1a8c361c26ae3251b21f505f9538336fdb14155) on Linux Mint 19.3 (x86_64):

Thank you for reporting this. I fixed the a similar problem on travis by installing the [libc6-dev-amd64-cross](https://packages.ubuntu.com/bionic/libc6-dev-amd64-cross) package, and maybe there is a similar package for mint. Either way, I'm going to try to reproduce the problem and eliminate the need for that dependency (details in #18915). IMO, it does not need to block this PR, but I will try to fix it soon anyway.

Sjors commented:
Concept ACK

Changes are simpler than the last time I reviewed #16367, but I didn't study the current build code in much detail.

I was unable to build depends for e1a8c361c26ae3251b21f505f9538336fdb14155 on macOS 10.15.4,:

```
make -C depends MULTIPROCESS=1
Building libmultiprocess...
[ 14%] Linking CXX static library libmultiprocess.a
Error running link command: No such file or directory
make[3]: *** [libmultiprocess.a] Error 2
make[2]: *** [CMakeFiles/multiprocess.dir/all] Error 2
make[1]: *** [all] Error 2
make: *** [/Users/sjors/dev/bitcoin-ipc/depends/work/build/x86_64-apple-darwin19.4.0/libmultiprocess/5741d750a04e644a03336090d8979c6d033e32c0-332cf4b0b73/./.stamp_built] Error 2
```

Compiling without depends works fine (installed the latest master version of libmultiprocess).

Starting tests with `BITCOIND` breaks the backwards compatibility test: `BITCOIND=bitcoin-node test/functional/feature_backwards_compatibility.py`

```
FileNotFoundError: [Errno 2] No such file or directory: '/Users/sjors/dev/bitcoin-ipc/releases/v0.19.0.1/bin/bitcoind'
```

ryanofsky commented:
> I was unable to build depends for [e1a8c36](https://github.com/bitcoin/bitcoin/commit/e1a8c361c26ae3251b21f505f9538336fdb14155) on macOS 10.15.4,:

Boy, everything is broken after the changes to support linux->mac cross compile. Googling I think I might need to manually pass the host_AR variable, or just entirely give up on passing cross-compile options when the builder platform is the same as the host platform.

Thanks for testing.

> Starting tests with `BITCOIND` breaks the backwards compatibility test: `BITCOIND=bitcoin-node test/functional/feature_backwards_compatibility.py`
> 
> ```
> FileNotFoundError: [Errno 2] No such file or directory: '/Users/sjors/dev/bitcoin-ipc/releases/v0.19.0.1/bin/bitcoind'
> ```

Will try to reproduce but if you happen to know more about the failure, it'd be helpful. Not clear why setting BITCOIND=bitcoin-node would cause `bitcoin-ipc/releases/v0.19.0.1/bin/bitcoind` not to exist. I wouldn't expect the BITCOIND variable to affect contents of a download

hebasto commented:
> re: [#18677 (comment)](https://github.com/bitcoin/bitcoin/pull/18677#issuecomment-625674673)
> 
> > Testing [e1a8c36](https://github.com/bitcoin/bitcoin/commit/e1a8c361c26ae3251b21f505f9538336fdb14155) on Linux Mint 19.3 (x86_64):
> 
> Thank you for reporting this. I fixed the a similar problem on travis by installing the [libc6-dev-amd64-cross](https://packages.ubuntu.com/bionic/libc6-dev-amd64-cross) package, and maybe there is a similar package for mint. Either way, I'm going to try to reproduce the problem and eliminate the need for that dependency (details in #18915). IMO, it does not need to block this PR, but I will try to fix it soon anyway.

Thanks, it works. Btw, Linux Mint is based on Ubuntu's code base.

hebasto commented:
Testing e1a8c361c26ae3251b21f505f9538336fdb14155 on Linux Mint 19.3 (x86_64):
```
$ make -C depends NO_QT=1 MULTIPROCESS=1
$ CONFIG_SITE=$PWD/depends/x86_64-pc-linux-gnu/share/config.site ./configure
$ make check
$ BITCOIND=bitcoin-node test/functional/wallet_multiwallet.py
2020-05-09T06:57:09.915000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_p9ahj4dt
2020-05-09T06:57:12.472000Z TestFramework (ERROR): Assertion failed
Traceback (most recent call last):
  File "/home/hebasto/GitHub/bitcoin/test/functional/test_framework/test_node.py", line 475, in assert_start_raises_init_error
    self.wait_for_rpc_connection()
  File "/home/hebasto/GitHub/bitcoin/test/functional/test_framework/test_node.py", line 220, in wait_for_rpc_connection
    'bitcoind exited with status {} during initialization'.format(self.process.returncode)))
test_framework.test_node.FailedToStartError: [node 0] bitcoind exited with status -11 during initialization

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/hebasto/GitHub/bitcoin/test/functional/test_framework/test_framework.py", line 114, in main
    self.run_test()
  File "test/functional/wallet_multiwallet.py", line 95, in run_test
    self.nodes[0].assert_start_raises_init_error(['-wallet=wallet.dat/bad'], exp_stderr, match=ErrorMatch.PARTIAL_REGEX)
  File "/home/hebasto/GitHub/bitcoin/test/functional/test_framework/test_node.py", line 489, in assert_start_raises_init_error
    'Expected message "{}" does not partially match stderr:\n"{}"'.format(expected_msg, stderr))
  File "/home/hebasto/GitHub/bitcoin/test/functional/test_framework/test_node.py", line 161, in _raise_assertion_error
    raise AssertionError(self._node_msg(msg))
AssertionError: [node 0] Expected message "boost::filesystem::create_directory:" does not partially match stderr:
""
2020-05-09T06:57:12.524000Z TestFramework (INFO): Stopping nodes
2020-05-09T06:57:12.525000Z TestFramework (WARNING): Not cleaning up dir /tmp/bitcoin_func_test_p9ahj4dt
2020-05-09T06:57:12.525000Z TestFramework (ERROR): Test failed. Test logging available at /tmp/bitcoin_func_test_p9ahj4dt/test_framework.log
2020-05-09T06:57:12.525000Z TestFramework (ERROR): Hint: Call /home/hebasto/GitHub/bitcoin/test/functional/combine_logs.py '/tmp/bitcoin_func_test_p9ahj4dt' to consolidate all logs
```

ryanofsky commented:
Updated e1a8c361c26ae3251b21f505f9538336fdb14155 -> 952ac003aba664204979579e6f0363033a07cff2 ([`pr/ipc-build2.6`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build2.6) -> [`pr/ipc-build2.7`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build2.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-build2.6..pr/ipc-build2.7)) building natively without sysroot when possible in light of https://github.com/bitcoin/bitcoin/issues/18915

---

I responded to all issues reported above but I couldn't reproduce some problems. It's difficult to follow the threads in this PR and know in advance which problems are related, but I want to suggest if anybody sees any problems or errors at all with this PR please just go to https://github.com/chaincodelabs/libmultiprocess/issues/new and make a new issue pasting the error and build commands, and please open different issues if there are different problems.

---

re: https://github.com/bitcoin/bitcoin/pull/18677#pullrequestreview-408292513

> I fixed the a similar problem on travis by installing the [libc6-dev-amd64-cross](https://packages.ubuntu.com/bionic/libc6-dev-amd64-cross) package, and maybe there is a similar package for mint

Now with the new update, it should no longer be necessary to install this package;

---

re: https://github.com/bitcoin/bitcoin/pull/18677#issuecomment-625988135

> I was unable to build depends for [e1a8c36](https://github.com/bitcoin/bitcoin/commit/e1a8c361c26ae3251b21f505f9538336fdb14155) on macOS 10.15.4,:

I think this should be fixed in the latest update which switches back to a native build.

---

re: https://github.com/bitcoin/bitcoin/pull/18677#issuecomment-625988135

> Starting tests with `BITCOIND` breaks the backwards compatibility test: `BITCOIND=bitcoin-node test/functional/feature_backwards_compatibility.py`

I tried to reproduce this, but didn't couldn't get the problem to happen. As mentioned in the earlier reply https://github.com/bitcoin/bitcoin/pull/18677#issuecomment-625996555, it's difficult to see how changes from this PR or from setting the BITCOIND variable could cause `v0.19.0.1/bin/bitcoind` binary not to be found. My attempt to reproduce looked like:

```
git checkout e1a8c361c26ae3251b21f505f9538336fdb14155
./autogen.sh
export PKG_CONFIG_PATH=/home/russ/src/libmultiprocess/prefix/lib/pkgconfig
./configure --disable-ccache --enable-multiprocess
make
contrib/devtools/previous_release.sh -b v0.15.2 v0.16.3 v0.17.1 v0.18.1 v0.19.0.1
BITCOIND=bitcoin-node test/functional/feature_backwards_compatibility.py
```

Would definitely encourage creating an issue so we can get to the bottom of this at https://github.com/chaincodelabs/libmultiprocess/issues/new if you still see this error.

---

re: https://github.com/bitcoin/bitcoin/pull/18677#issuecomment-626118471

> ```
> test_framework.test_node.FailedToStartError: [node 0] bitcoind exited with status -11 during initialization
> ```

I tried to follow the same steps to reproduce but couldn't get this issue to happen. There's a chance this could be fixed in the latest update if there was some inconsitency in the depends sysroots causing this error. For for this error to happen at this point in the test is pretty mysterious given node would have had to start successfully before this. Would encourage creating an issue at https://github.com/chaincodelabs/libmultiprocess/issues/new if this error still happens or if you see any new errors from this PR. Here are steps I tried to reproduce this with:

```
git checkout e1a8c361c26ae3251b21f505f9538336fdb14155
make -C depends MULTIPROCESS=1 NO_QT=1 V=1
./autogen.sh
CONFIG_SITE=$PWD/depends/x86_64-pc-linux-gnu/share/config.site ./configure --disable-ccache
make
BITCOIND=bitcoin-node test/functional/wallet_multiwallet.py
```

MarcoFalke commented on .travis.yml:
```diff
@@ -148,6 +148,11 @@ jobs:
       env: >-
         FILE_ENV="./ci/test/00_setup_env_native_fuzz.sh"
 
+    - stage: test
```
not sure if this is useful to run on travis. This is an experimental feature (currently not even a feature). So the inconvenience in the unlikely case that it breaks is probably not worth having it run on every pull request. Please keep the ci config, though. If you want you can add it to `.cirrus.ci` here in the repo or people can run it manually on their machines.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/18677#discussion_r423688995

> not sure if this is useful to run on travis

What if I added an `if: commit_message !~ /depends:|multiprocess:/` [condition](https://docs.travis-ci.com/user/conditional-builds-stages-jobs)? This travis config has uncovered a few build and runtime bugs that weren't happening for me locally, so I think it's been pretty useful. The idea behind the regex would be not run on normal pull requests but catch regressions on depends and multiprocess pull requests.

MarcoFalke replied:
Absolutely, if you can get the condition to work. Travis is running the build on a temporary merge commit, so the commit message might not might not be the one you pushed.

ryanofsky commented:
Rebased 952ac003aba664204979579e6f0363033a07cff2 -> c4378e6bafcf1d749293519ab9dde10d8b6d2230 ([`pr/ipc-build2.7`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build2.7) -> [`pr/ipc-build2.8`](https://github.com/ryanofsky/bitcoin/commits/pr/ipc-build2.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/ipc-build2.7-rebase..pr/ipc-build2.8)) due to conflict with #18929. Also added condition to skip travis multiprocess depends build on non-depends, non-multiprocess PRs

hebasto reviewed (approved):
ACK e2bab2aa162ae38b2bf8195b577c982402fbee9d, tested on Linux Mint 19.3 (x86_64):
```
$ make -C depends NO_QT=1 MULTIPROCESS=1
$ ./autogen.sh
$ CONFIG_SITE=$PWD/depends/x86_64-pc-linux-gnu/share/config.site ./configure
$ make check
$ BITCOIND=bitcoin-node ./test/functional/test_runner.py
```

Also built depends without `NO_QT=1` and tested `./src/bitcoin-gui` :+1:

Sjors reviewed (approved):
tACK e2bab2aa162ae38b2bf8195b577c982402fbee9d on macOS 10.15.4

Depends builds now. Without depends still works fine too, using `capnp` 0.7.0 from homebrew, and the latest `libmultiprocess` self compiled. Ideally you should tag a `libmultiprocess` (pre-)release after this lands in master, or right before 0.21 branches off.

I'm no longer able to reproduce a problem with `BITCOIND=bitcoin-node test/functional/feature_backwards_compatibility.py`

Direct link to IRC meeting log for April 16th: http://www.erisian.com.au/meetbot/bitcoin-core-dev/2020/bitcoin-core-dev.2020-04-16-19.03.log.html#l-243 

>> I thought we were going to pivot this to just add a `-process=node/wallet/gui` option to the same `bitcoin-qt` binary?

>  think the current naming is better, but I pointed out in the IRC meeting it is easy to change. 

I think separate binaries is a better approach for now. For development I'd like to have them both, so I don't have to reconfigure to see if a bug is related to the multiprocess stuff. I imagine at some point we'll feel comfortable shipping the multi-process binaries, with their dependencies, but not comfortable enough to make them the default.

ryanofsky commented on depends/packages/native_capnp.mk:
```diff
@@ -0,0 +1,18 @@
+$(package)_download_file=capnproto-c++-$($(package)_version).tar.gz
+$(package)_file_name=capnproto-cxx-$($(package)_version).tar.gz
+$(package)_sha256_hash=c9a4c0bd88123064d483ab46ecee777f14d933359e23bff6fb4f4dbd28b4cd41
+
```
re: https://github.com/bitcoin/bitcoin/pull/18677#discussion_r427254248

> It's also consistent with the practice in other dependencies to actively opt out of stuff (see e.g. #17730, #16370)

Thanks, I will make that change in a followup, or here if another update is required (reluctant to make changes now it seems like all the build problems so far have been ironed out).

At the time I made that comment about OpenSSL I mistakenly thought the depends build was more hermetic, and that it wouldn't pick up system dependencies in host package builds. This was before I knew about #18915

ryanofsky commented:
I think it might be a good time to merge this PR. It seems to be at a point where it is working for the people who've tested it on different platforms. There are definitely improvements to be made going forward, but we know build changes are fragile and a fix for one bug tends to lead to other bugs, so it would be great to be done with this big PR when it's in a pretty good state, and start to track future issues in smaller more focused PRs.

Sjors commented:
Maybe a Gitian build to top it off?

fanquake commented:
> I think it might be a good time to merge this PR.

Thanks. I'm planning on merging this shortly.

practicalswift commented:
ACK e2bab2aa162ae38b2bf8195b577c982402fbee9d

Thanks for doing this excellent architectural work: a dream come true (even if optional)!

With this dream off the table I think I only have [Sandboxed API integration](https://developers.google.com/sandboxed-api), [control-flow integrity](https://clang.llvm.org/docs/ControlFlowIntegrity.html) and killing the UUM bug class for good using [`-ftrivial-auto-var-init=zero`](https://reviews.llvm.org/D54604?id=174471) left to dream about :)

Thanks @ryanofsky: you're great, and so is your patience -- this has been a very long road :)

MarcoFalke commented:
Checked that the ci config file is indeed native and runs on aarch64:

```
Options used to compile and link:
  multiprocess  = yes
  with wallet   = yes
  with gui / qt = yes
    with qr     = yes
  with zmq      = yes
  with test     = yes
    with fuzz   = no
  with bench    = yes
  with upnp     = yes
  use asm       = yes
  sanitizers    = 
  debug enabled = no
  gprof enabled = no
  werror        = no

  target os     = linux
  build os      = 

  CC            = /usr/bin/ccache gcc
  CFLAGS        = -pipe -O2 
  CPPFLAGS      =   -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=2 -I/bitcoin-core/depends/aarch64-unknown-linux-gnu/share/../include/  -DHAVE_BUILD_INFO -D__STDC_FORMAT_MACROS
  CXX           = /usr/bin/ccache g++ -std=c++11
  CXXFLAGS      =   -fstack-reuse=none -Wstack-protector -fstack-protector-all     -pipe -O2 
  LDFLAGS       = -pthread  -Wl,-z,relro -Wl,-z,now -pie  -L/bitcoin-core/depends/aarch64-unknown-linux-gnu/share/../lib 
  ARFLAGS       = cr
